docs/source/jit_type_system_1.rst

```diff
@@ -0,0 +1,628 @@
+
+Since the type schema of module instance class is constructed directly from an instance object (created outside the scope of TorchScript), rather than inferred from ``__init__()`` like custom classes. It is possible that two objects of the same instance class type follow two different type schemas.
+
+In this sense, ``TSModuleType`` is not really a static type. Therefore, for type safety considerations, ``TSModuleType`` cannot be used in TorchScript type annotation or be composed with ``TSType``.
+
+Module Instance Class
+^^^^^^^^^^^^^^^^^^^^^
+
+TorchScript module type represents type schema of a user-defined PyTorch module instance.  When scripting a PyTorch module, the module object is always created outside TorchScript (i.e., passed in as parameter to forward()), the Python module class is treated as a module instance class so that ``__init__()`` of the Python module class are not subject to the type checking rules of TorchScript
+
+::
+
+    TSModuleType := "class" Identifier "(torch.nn.Module)" ":"
+                        ClassBodyDefinition
+
+where
+
+* ``forward()`` and other methods decorated with ``@torch.jit.export`` must be compilable to TorchScript IR and subject to TorchScript’s type checking rules
+
+Unlike custom classes, only the forward method and other methods decorated with ``@torch.jit.export``  of the module type need to be compilable to TorchScript IR. Most notably, ``__init__()`` is not considered a TorchScript method. Thus module type constructors should not be invoked within the scope of TorchScript. Instead, TorchScript module objects are always constructed outside and passed into ``torch.jit.script(ModuleObj)``.
+
+Example:
+
+.. testcode::
+
+    import torch
+    class TestModule(torch.nn.Module):
+        def __init__(self, v):
+            super().__init__()
+            self.x = v
+
+        def mul(self, x: int):
+            return self.x * x
+
+        @torch.jit.export
+        def madd(self, x: int, y: int):
+            return self.mul(x) + y
+
+        def forward(self, x: int):
+            return self.x + x
+
+    m = torch.jit.script(TestModule(2))
+    print(m.madd(2, 3))
+
+    m = torch.jit.script(TestModule(2.0))
+    print(m.madd(2, 3))
+
+.. testoutput::
+
+    7
+    7.0
+
+This example illustrates a few features of module types:
+
+*  The TestModule instance is created outside the scope of TorchScript (i.e., before invoking torch.jit.script).
+* ``__init__()`` is not considered as a TorchScript method, therefore it does not have to be annotated and can contain arbitrary Python codes. In fact, it is prohibited to invoke ``__init__()`` of a instance class within the scope of TorchScript execution.
+* Because TestModule instances are instantiated in Python, in this example, TestModule(2.0) and TestModule(2) create two instances with different types for its data attributes, i.e., (self.x is of type float for TestModule(2.0), whereas self.y is of type int for TestModule(2.0)).
```

Comment:
```suggestion
* Because ``TestModule`` instances are instantiated in Python, in this example, ``TestModule(2.0)`` and ``TestModule(2)`` create two instances with different types for its data attributes. ``self.x is of type ``float`` for ``TestModule(2.0)``, whereas ``self.y`` is of type ``int`` for ``TestModule(2.0)``.
```